cf-custom-resources/lib/cert-replicator.js

```diff
@@ -0,0 +1,320 @@
+          resolve();
+        }
+      })
+      .end(responseBody, "utf8");
+  });
+};
+
+/**
+ * Replicate a public certificate from AWS Certificate Manager in the target region.
+ *
+ * @param {string} requestId the CloudFormation request ID
+ * @param {string} appName the application name
+ * @param {string} envName the environment name
+ * @param {string} certArn arn for the certificate to replicate from
+ * @param {string} envRegionAcm acm client in environment region
+ * @param {string} targetRegionAcm acm client in target region
+ * @returns {string} ARN of the replicated certificate
+ */
+const replicateCertificate = async function (
+  requestId,
+  appName,
+  envName,
+  certArn,
+  envRegionAcm,
+  targetRegionAcm
+) {
+  const { Certificate } = await envRegionAcm
+    .describeCertificate({
+      CertificateArn: certArn,
+    })
+    .promise();
+  const domainName = Certificate.DomainName;
+  const sans = Certificate.SubjectAlternativeNames;
+
+  const crypto = require("crypto");
+  return targetRegionAcm
+    .requestCertificate({
+      DomainName: domainName,
+      SubjectAlternativeNames: sans,
+      IdempotencyToken: crypto
+        .createHash("sha256")
+        .update(requestId)
+        .digest("hex")
+        .substr(0, 32),
+      ValidationMethod: "DNS",
+      Tags: [
+        {
+          Key: "copilot-application",
+          Value: appName,
+        },
+        {
+          Key: "copilot-environment",
+          Value: envName,
+        },
+      ],
+    })
+    .promise();
+};
+
+/**
+ * Deletes a certificate from AWS Certificate Manager (ACM) by its ARN.
+ * Specifically, if it is the last certificate attaching to the listener, it will also remove the CNAME records
+ * for validation in all the root, app, and env hosted zones in parallel.
+ * If the certificate does not exist, the function will return normally.
+ *
+ * @param {string} arn The certificate ARN
+ * @param {string} targetRegionAcm acm client in target region
+ */
+const deleteCertificate = async function (arn, acm) {
+  try {
+    let inUseByResources;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      const { Certificate } = await acm
+        .describeCertificate({
+          CertificateArn: arn,
+        })
+        .promise();
+
+      inUseByResources = Certificate.InUseBy || [];
+      if (!inUseByResources.length) {
+        break;
+      }
+      // Deleting resources can be quite slow - so just sleep 30 seconds between checks.
+      await sleep(30000);
+    }
+    if (inUseByResources.length) {
+      throw new Error(
+        `Certificate still in use by ${inUseByResources.join()} after checking for ${maxAttempts} attempts.`
+      );
+    }
+
+    await acm
+      .deleteCertificate({
+        CertificateArn: arn,
+      })
+      .promise();
+  } catch (err) {
+    if (err.name !== "ResourceNotFoundException") {
+      throw err;
+    }
+  }
+};
+
+const validateCertificate = async function (certificateARN, acm) {
```

Comment:
How are the validation records upserted? Or do we not need to because the original cert's validation records would be the same as this one, meaning that this replicated cert has already been validated before it was created?

Reply:
Yeah that's true.